Redis module doesn't connect when using native image

the api microservice won't connect to redis

```
08:11:00.617 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Error instantiating bean of type [io.lettuce.core.pubsub.StatefulRedisPubSubConnection]: io.lettuce.core.pubsub.api.sync.RedisPubSubCommands.dispatch(io.lettuce.core.protocol.ProtocolKeyword, io.lettuce.core.output.CommandOutput, io.lettuce.core.protocol.CommandArgs)
io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type [io.lettuce.core.pubsub.StatefulRedisPubSubConnection]: io.lettuce.core.pubsub.api.sync.RedisPubSubCommands.dispatch(io.lettuce.core.protocol.ProtocolKeyword, io.lettuce.core.output.CommandOutput, io.lettuce.core.protocol.CommandArgs)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2008)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2770)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2756)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2427)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2401)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:1285)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:755)
	at io.micronaut.context.BeanLocator.findBean(BeanLocator.java:149)
	at io.micronaut.configuration.lettuce.session.RedisSessionStore.lambda$findRedisPubSubConnection$22(RedisSessionStore.java:405)
	at java.util.Optional.orElseGet(Optional.java:369)
	at io.micronaut.configuration.lettuce.session.RedisSessionStore.findRedisPubSubConnection(RedisSessionStore.java:405)
	at io.micronaut.configuration.lettuce.session.RedisSessionStore.<init>(RedisSessionStore.java:142)
	at io.micronaut.configuration.lettuce.session.$RedisSessionStoreDefinition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1979)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2770)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2756)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2427)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2401)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1264)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1014)
	at io.micronaut.session.binder.$SessionArgumentBinderDefinition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1979)
	at io.micronaut.context.DefaultBeanContext.addCandidateToList(DefaultBeanContext.java:3141)
	at io.micronaut.context.DefaultBeanContext.getBeanRegistrations(DefaultBeanContext.java:3044)
	at io.micronaut.context.DefaultBeanContext.getBeansOfType(DefaultBeanContext.java:1062)
	at io.micronaut.context.AbstractBeanDefinition.doGetBeansOfType(AbstractBeanDefinition.java:1232)
	at io.micronaut.context.AbstractBeanDefinition.getBeansOfTypeForConstructorArgument(AbstractBeanDefinition.java:1179)
	at io.micronaut.http.bind.$DefaultRequestBinderRegistryDefinition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1979)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2770)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2756)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2427)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2401)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1264)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1014)
	at io.micronaut.http.server.netty.$NettyRequestArgumentSatisfierDefinition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1979)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2770)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2756)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2427)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2401)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1264)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1014)
	at io.micronaut.http.server.netty.$NettyHttpServerDefinition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1979)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:2770)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:2756)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2427)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2401)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:1285)
	at io.micronaut.context.DefaultBeanContext.findBean(DefaultBeanContext.java:755)
	at io.micronaut.context.BeanLocator.findBean(BeanLocator.java:149)
	at io.micronaut.runtime.Micronaut.start(Micronaut.java:73)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:311)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:297)
	at api.Application.main(Application.java:25)
Caused by: java.lang.NoSuchMethodError: io.lettuce.core.pubsub.api.sync.RedisPubSubCommands.dispatch(io.lettuce.core.protocol.ProtocolKeyword, io.lettuce.core.output.CommandOutput, io.lettuce.core.protocol.CommandArgs)
	at com.sun.proxy.$Proxy345.<clinit>(Unknown Source)
	at com.oracle.svm.core.classinitialization.ClassInitializationInfo.invokeClassInitializer(ClassInitializationInfo.java:375)
	at com.oracle.svm.core.classinitialization.ClassInitializationInfo.initialize(ClassInitializationInfo.java:295)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.lang.reflect.Proxy.newProxyInstance(Proxy.java:1022)
	at java.lang.reflect.Proxy.newProxyInstance(Proxy.java:1008)
	at io.lettuce.core.RedisChannelHandler.syncHandler(RedisChannelHandler.java:314)
	at io.lettuce.core.pubsub.StatefulRedisPubSubConnectionImpl.newRedisSyncCommandsImpl(StatefulRedisPubSubConnectionImpl.java:103)
	at io.lettuce.core.pubsub.StatefulRedisPubSubConnectionImpl.newRedisSyncCommandsImpl(StatefulRedisPubSubConnectionImpl.java:45)
	at io.lettuce.core.StatefulRedisConnectionImpl.<init>(StatefulRedisConnectionImpl.java:81)
	at io.lettuce.core.pubsub.StatefulRedisPubSubConnectionImpl.<init>(StatefulRedisPubSubConnectionImpl.java:61)
	at io.lettuce.core.RedisClient.newStatefulRedisPubSubConnection(RedisClient.java:626)
	at io.lettuce.core.RedisClient.connectPubSubAsync(RedisClient.java:405)
	at io.lettuce.core.RedisClient.connectPubSub(RedisClient.java:328)
	at io.micronaut.configuration.lettuce.AbstractRedisClientFactory.redisPubSubConnection(AbstractRedisClientFactory.java:80)
	at io.micronaut.configuration.lettuce.DefaultRedisClientFactory.redisPubSubConnection(DefaultRedisClientFactory.java:62)
	at io.micronaut.configuration.lettuce.$DefaultRedisClientFactory$RedisPubSubConnection2Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:1979)
```

--- a/.github/workflows/graalvm.yml
+++ b/.github/workflows/graalvm.yml
@@ -19,7 +19,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        java: ['java11', 'java16']
+        java: ['java11']
     steps:
        # https://github.com/actions/virtual-environments/issues/709
       - name: Free disk space

--- a/.github/workflows/graalvm.yml
+++ b/.github/workflows/graalvm.yml
@@ -19,7 +19,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        graalvm: ['21.0.0.2.java8', '21.0.0.2.java11']
+        java: ['java11', 'java16']
     steps:
        # https://github.com/actions/virtual-environments/issues/709
       - name: Free disk space
@@ -38,7 +38,8 @@ jobs:
       - name: Setup GraalVM CE
         uses: DeLaGuardo/setup-graalvm@4.0
         with:
-          graalvm-version: ${{ matrix.graalvm }}
+          graalvm: 21.1.0
+          java: ${{ matrix.java }}
       - name: Install Native Image
         run: gu install native-image
       - name: Build with Gradle

--- a/build.gradle
+++ b/build.gradle
@@ -9,6 +9,11 @@ buildscript {
 }
 
 subprojects { Project subproject ->
+
+    if (subproject.path.contains("tests")) {
+        return
+    }
+
     group "io.micronaut.redis"
 
     apply plugin: "io.micronaut.build.internal.common"

--- a/gradle.properties
+++ b/gradle.properties
@@ -5,6 +5,8 @@ micronautTestVersion=2.3.3
 micronautBuildVersion=1.1.5
 groovyVersion=3.0.8
 spockVersion=2.0-M5-groovy-3.0
+micronautGradlePluginVersion=1.5.0
+testContainersVersion=1.15.3
 
 lettuceVersion=6.1.1.RELEASE
 micrometerVersion=1.5.1

--- a/redis-lettuce/build.gradle
+++ b/redis-lettuce/build.gradle
@@ -2,7 +2,8 @@ dependencies {
     annotationProcessor "io.micronaut:micronaut-inject-java"
     // needed because otherwise IntelliJ integration broken
     compileOnly "io.micronaut:micronaut-inject-java"
-    annotationProcessor "io.micronaut.docs:micronaut-docs-asciidoc-config-props:$micronautDocsVersion"    
+    annotationProcessor "io.micronaut.docs:micronaut-docs-asciidoc-config-props:$micronautDocsVersion"
+    annotationProcessor "io.micronaut:micronaut-graal"
 
     api "io.lettuce:lettuce-core:$lettuceVersion"
     api "io.micronaut.cache:micronaut-cache-core"

--- a/redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/session/RedisSessionStore.java
+++ b/redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/session/RedisSessionStore.java
@@ -25,6 +25,8 @@ import io.lettuce.core.cluster.api.StatefulRedisClusterConnection;
 import io.lettuce.core.cluster.api.async.RedisAdvancedClusterAsyncCommands;
 import io.lettuce.core.cluster.api.sync.RedisAdvancedClusterCommands;
 import io.lettuce.core.pubsub.RedisPubSubAdapter;
+import io.lettuce.core.pubsub.RedisPubSubAsyncCommandsImpl;
+import io.lettuce.core.pubsub.RedisPubSubReactiveCommandsImpl;
 import io.lettuce.core.pubsub.StatefulRedisPubSubConnection;
 import io.lettuce.core.pubsub.api.sync.RedisPubSubCommands;
 import io.micronaut.configuration.lettuce.RedisConnectionUtil;
@@ -34,6 +36,7 @@ import io.micronaut.context.annotation.Replaces;
 import io.micronaut.context.annotation.Requires;
 import io.micronaut.context.event.ApplicationEventPublisher;
 import io.micronaut.context.exceptions.ConfigurationException;
+import io.micronaut.core.annotation.TypeHint;
 import io.micronaut.core.convert.ArgumentConversionContext;
 import io.micronaut.core.convert.value.MutableConvertibleValues;
 import io.micronaut.core.serialize.ObjectSerializer;
@@ -92,6 +95,7 @@ import static io.micronaut.configuration.lettuce.session.RedisSessionStore.Redis
 @Primary
 @Requires(property = RedisSessionStore.REDIS_SESSION_ENABLED, value = StringUtils.TRUE)
 @Replaces(InMemorySessionStore.class)
+@TypeHint(value = {RedisPubSubAsyncCommandsImpl.class, RedisPubSubReactiveCommandsImpl.class}, accessType = TypeHint.AccessType.ALL_PUBLIC)
 public class RedisSessionStore extends RedisPubSubAdapter<String, String> implements SessionStore<RedisSessionStore.RedisSession>, AutoCloseable {
 
     public static final String REDIS_SESSION_ENABLED = SessionSettings.HTTP + ".redis.enabled";

--- a/settings.gradle
+++ b/settings.gradle
@@ -1,4 +1,12 @@
+pluginManagement {
+    plugins {
+        id 'io.micronaut.application' version getProperty("micronautGradlePluginVersion")
+        id 'io.micronaut.library' version getProperty("micronautGradlePluginVersion")
+    }
+}
+
 rootProject.name = 'redis'
 
 include 'redis-lettuce'
+include 'tests:http-session'
 

--- a/tests/http-session/build.gradle
+++ b/tests/http-session/build.gradle
@@ -1,0 +1,28 @@
+plugins {
+    id "java"
+    id "groovy"
+    id "io.micronaut.application"
+}
+
+application {
+    mainClass = "example.Application"
+}
+
+micronaut {
+    version project.ext.micronautVersion
+    runtime "netty"
+    testRuntime "spock"
+}
+
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    implementation project(":redis-lettuce")
+    implementation "io.micronaut:micronaut-session"
+    testImplementation "io.micronaut:micronaut-http-client"
+    testImplementation platform("org.testcontainers:testcontainers-bom:$testContainersVersion")
+    testImplementation "org.testcontainers:testcontainers"
+    testImplementation "org.testcontainers:spock"
+}

--- a/tests/http-session/src/main/java/example/Application.java
+++ b/tests/http-session/src/main/java/example/Application.java
@@ -1,0 +1,11 @@
+
+package example;
+
+import io.micronaut.runtime.Micronaut;
+
+public class Application {
+
+    public static void main(String[] args) {
+        Micronaut.run(Application.class);
+    }
+}

--- a/tests/http-session/src/main/java/example/SessionController.java
+++ b/tests/http-session/src/main/java/example/SessionController.java
@@ -1,0 +1,18 @@
+package example;
+
+import io.micronaut.http.annotation.Controller;
+import io.micronaut.http.annotation.Get;
+import io.micronaut.session.Session;
+
+@Controller("/sessions")
+public class SessionController {
+
+    @Get
+    String simple(Session session) {
+        return (String) session.get("myValue").orElseGet(() -> {
+            session.put("myValue", "value in session");
+            return "not in session";
+        });
+    }
+
+}

--- a/tests/http-session/src/main/resources/logback.xml
+++ b/tests/http-session/src/main/resources/logback.xml
@@ -1,0 +1,17 @@
+<configuration>
+
+    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
+        <withJansi>true</withJansi>
+        <!-- encoders are assigned the type
+             ch.qos.logback.classic.encoder.PatternLayoutEncoder by default -->
+        <encoder>
+            <pattern>%cyan(%d{HH:mm:ss.SSS}) %gray([%thread]) %highlight(%-5level) %magenta(%logger{36}) - %msg%n</pattern>
+        </encoder>
+    </appender>
+
+    <root level="info">
+        <appender-ref ref="STDOUT" />
+    </root>
+
+
+</configuration>

--- a/tests/http-session/src/test/groovy/example/SessionSpec.groovy
+++ b/tests/http-session/src/test/groovy/example/SessionSpec.groovy
@@ -1,0 +1,59 @@
+package example
+
+import io.micronaut.context.annotation.Property
+import io.micronaut.http.HttpHeaders
+import io.micronaut.http.HttpRequest
+import io.micronaut.http.client.HttpClient
+import io.micronaut.http.client.annotation.Client
+import io.micronaut.test.extensions.spock.annotation.MicronautTest
+import io.micronaut.test.support.TestPropertyProvider
+import org.testcontainers.containers.GenericContainer
+import org.testcontainers.utility.DockerImageName
+import spock.lang.AutoCleanup
+import spock.lang.Shared
+import spock.lang.Specification
+
+import javax.inject.Inject
+
+@MicronautTest
+@Property(name = "micronaut.session.http.redis.enabled", value = "true")
+class SessionSpec extends Specification implements TestPropertyProvider {
+
+    @Shared
+    @AutoCleanup
+    GenericContainer redis = new GenericContainer(DockerImageName.parse("redis:5.0.3-alpine"))
+            .withExposedPorts(6379)
+
+    @Inject
+    @Client("/")
+    HttpClient client
+
+    void "test session"() {
+        when:
+            def response = client.toBlocking().exchange(
+                    HttpRequest.GET("/sessions"), String
+            )
+
+        then:
+            response.getBody().get() == "not in session"
+            response.header(HttpHeaders.SET_COOKIE)
+
+        when:
+            def sessionId = response.header(HttpHeaders.SET_COOKIE)
+            response = client.toBlocking().exchange(
+                    HttpRequest.GET("/sessions").header(HttpHeaders.COOKIE, sessionId), String
+            )
+
+        then:
+            response.getBody().get() == "value in session"
+            response.header(HttpHeaders.SET_COOKIE)
+    }
+
+    @Override
+    Map<String, String> getProperties() {
+        redis.start()
+        return [
+                'redis.uri': 'redis://' + redis.getContainerIpAddress() + ":" + redis.getMappedPort(6379)
+        ]
+    }
+}

